srcRs/DustDDS/src/main.rs

```diff
@@ -0,0 +1,737 @@
+                    default_color
+                );
+                default_color
+            }
+        }
+    }
+
+    fn reliability_qos_policy(&self) -> ReliabilityQosPolicy {
+        let mut reliability = DataWriterQos::default().reliability;
+        if self.best_effort_reliability {
+            reliability.kind = qos_policy::ReliabilityQosPolicyKind::BestEffort;
+        }
+        if self.reliable_reliability {
+            reliability.kind = qos_policy::ReliabilityQosPolicyKind::Reliable;
+        }
+        reliability
+    }
+
+    fn partition_qos_policy(&self) -> PartitionQosPolicy {
+        if let Some(partition) = &self.partition {
+            PartitionQosPolicy {
+                name: vec![partition.to_owned()],
+            }
+        } else {
+            PartitionQosPolicy::default()
+        }
+    }
+
+    fn durability_qos_policy(&self) -> DurabilityQosPolicy {
+        DurabilityQosPolicy {
+            kind: match self.durability_kind {
+                'v' => qos_policy::DurabilityQosPolicyKind::Volatile,
+                'l' => qos_policy::DurabilityQosPolicyKind::TransientLocal,
+                't' => qos_policy::DurabilityQosPolicyKind::Transient,
+                'p' => qos_policy::DurabilityQosPolicyKind::Persistent,
+                _ => panic!("durability not valid"),
+            },
+        }
+    }
+
+    fn data_representation_qos_policy(&self) -> DataRepresentationQosPolicy {
+        let data_representation = match self.data_representation {
+            1 => XCDR_DATA_REPRESENTATION,
+            2 => XCDR2_DATA_REPRESENTATION,
+            _ => panic!("Wrong data representation"),
+        };
+        qos_policy::DataRepresentationQosPolicy {
+            value: vec![data_representation],
+        }
+    }
+
+    fn ownership_qos_policy(&self) -> OwnershipQosPolicy {
+        OwnershipQosPolicy {
+            kind: match self.ownership_strength {
+                -1 => qos_policy::OwnershipQosPolicyKind::Shared,
+                _ => qos_policy::OwnershipQosPolicyKind::Exclusive,
+            },
+        }
+    }
+
+    fn history_depth_qos_policy(&self) -> HistoryQosPolicy {
+        match self.history_depth {
+            -1 => HistoryQosPolicy::default(),
+            0 => HistoryQosPolicy {
+                kind: HistoryQosPolicyKind::KeepAll,
+            },
+            x if x >= 1 => HistoryQosPolicy {
+                kind: HistoryQosPolicyKind::KeepLast(x as u32),
+            },
+            _ => panic!("history_depth not valid"),
+        }
+    }
+
+    fn ownership_strength_qos_policy(&self) -> OwnershipStrengthQosPolicy {
+        if self.ownership_strength < -1 {
+            panic!("Ownership strength must be positive or zero")
+        }
+        OwnershipStrengthQosPolicy {
+            value: self.ownership_strength,
+        }
+    }
+}
+
+#[derive(Debug, dust_dds::topic_definition::type_support::DdsType)]
+pub struct ShapeType {

```

Comment:
Is there a way of providing the ShapeType as an IDL file?

Reply:
My knowledge about rust is limited, so I only have one comment about the type definition

Reply:
That is actually possible. We could use the idl that is already present in the srcCxx folder.